### PR TITLE
Auth/PM-12613 - Registration with Email Verification - Provider Invite Flow

### PR DESCRIPTION
**Pull request by JaredSnider-Bitwarden**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-12613
Clients PR: https://github.com/bitwarden/clients/pull/11635

## 📔 Objective

To add support for the provider invite flow in the new registration with email verification flow.

## 📸 Screenshots

See clients PR.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

**Review by ike-kottlowski — COMMENTED** [review]
💭 : Could we move the logic from the `AccountController.PostRegisterFinish([FromBody] RegisterFinishRequestModel model)` into `RegisterFinishRequestModel` object?

We could add a `RegisterFinishRequestModel.GetType()` method that would do all the logic we are doing in the account controller but within the object allowing the object to manage itself.

We would then need to make an RegisterFinishInviteType Enum.

``` C#
namespace Bit.Core.Auth.Enums;

public enum RegisterFinishInviteType : byte
{
    OrganizationInvite = 0,
    OrganizationSponsoredFreeFamilyPlanInvite = 1,
    EmercencyAccessInvite = 2,
    ProviderInvite = 3,
    EmailVerification = 4,
}
```

The `RegisterFinishRequestModel.GetType()` method would be similar to the logic in the accounts controller.

``` C#
public RegisterFinishInviteType GetType()
{
    ...
    if (!string.IsNullOrEmpty(model.ProviderInviteToken) && model.ProviderUserId.HasValue)
    {
      return RegisterFinishINviteType.ProviderInvite;
    }
    ...
}
```

Then in the `AccountController.PostRegisterFinish(...)` method we could use a `switch`
``` C#
public async Task<RegisterResponseModel> PostRegisterFinish([FromBody] RegisterFinishRequestModel model)
{
  ...
  Switch(RegisterFinishRequestModel.GetType())
  {
    ...
    case(ProviderInvite):
      identityResult = await _registerUserCommand.RegisterUserViaProviderInviteToken(user, model.MasterPasswordHash,
        model.ProviderInviteToken, model.ProviderUserId.Value);
      break;
    ...
    default:
      throw new BadRequestException("Invalid registration finish request");
  }
  return await ProcessRegistrationResult(identityResult, user, delaysEnabled);
}
```

I like the idea of the controller executing the command. I think the account controller should be able to ask the object what it's trying to do instead of figuring it out. It's smoke and mirrors but I think it does better encapsulation. 

The change has net zero impact on function, it could be teed-up to tech-debt. 🤷🏻

File: src/Core/Auth/UserFeatures/Registration/IRegisterUserCommand.cs [hunk 1]
```diff
@@ -61,4 +61,17 @@ public interface IRegisterUserCommand
     public Task<IdentityResult> RegisterUserViaAcceptEmergencyAccessInviteToken(User user, string masterPasswordHash,
         string acceptEmergencyAccessInviteToken, Guid acceptEmergencyAccessId);
 
+    // add summary
```
[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply src/Core/Auth/UserFeatures/Registration/IRegisterUserCommand.cs:64]: Ah, good catch.

**Comment by JaredSnider-Bitwarden** [comment]
> ut I think it does better encapsulation.
> 
> The change has net zero impact on function, it could be teed-up to tech-debt. 🤷🏻

@ike-kottlowski , thanks for the great idea. I've implemented the feedback. The integration tests all passed + I retested all the flows manually as well to double check!

**Comment by JaredSnider-Bitwarden** [comment]
> @ike-kottlowski , thanks for the great idea. I've implemented the feedback. The integration tests all passed + I retested all the flows manually as well to double check!

File: src/Core/Utilities/CoreHelpers.cs
Comment (ike-kottlowski) [inline src/Core/Utilities/CoreHelpers.cs:1]: ❓ : is the `Console.WriteLines()` meant to be here?

File: src/Core/Auth/Models/Api/Request/Accounts/RegisterFinishRequestModel.cs
Comment (ike-kottlowski) [inline src/Core/Auth/Models/Api/Request/Accounts/RegisterFinishRequestModel.cs:1]: 💭 : Is it possible to create Unit tests to test the `GetTokenType()` method? 

Something like [this](https://github.com/bitwarden/server/blob/a952d106374d838b1b526b3289a8d152240c1904/test/Api.Test/Auth/Models/Response/UserTwoFactorDuoResponseModelTests.cs#L9)?

File: src/Identity/Controllers/AccountsController.cs
Comment (ike-kottlowski) [inline src/Identity/Controllers/AccountsController.cs:1]: ❤️ : Looks great!

**Review by ike-kottlowski — COMMENTED** [review]
Looks great!

Hoping to get some unit tests for the new `GetTokenType()` method.

File: src/Core/Utilities/CoreHelpers.cs
Comment (JaredSnider-Bitwarden) [reply src/Core/Utilities/CoreHelpers.cs:1]: Argh, great catch. Removed.

File: src/Core/Auth/Models/Api/Request/Accounts/RegisterFinishRequestModel.cs
Comment (JaredSnider-Bitwarden) [reply src/Core/Auth/Models/Api/Request/Accounts/RegisterFinishRequestModel.cs:1]: Great idea. Tests added.

**Review by ike-kottlowski — APPROVED** [review]

**Pull request merged**